lib/ansible/modules/cloud/amazon/ec2_group.py

```diff
@@ -943,7 +943,7 @@ def get_final_rules(client, module, security_group_rules, specified_rules, purge
                         format_rule['user_id_group_pairs'][0].pop(k)
             final_rules.append(format_rule)
             # Order final rules consistently
-            final_rules.sort(key=lambda x: x.get('cidr_ip', x.get('ip_ranges', x.get('ipv6_ranges', x.get('prefix_list_ids', x.get('user_id_group_pairs'))))))
+            final_rules.sort(key=lambda x: json.dumps(x, sort_keys=True))
```

Comment:
It looks like this prevents sorting to happen at all, because the string returned by json.dumps is not in final_results. I think the initial code was trying to sort the final rules according to a defining rule characteristic - `cidr_ip`, `ip_ranges`, `ipv6_ranges`, `prefix_list_ids`, and `user_id_group_pairs`, in that order.
The existing code isn't very readable and I don't think it does what it's trying to do (are cidr_ip or ipv6_ranges ever top level keys? For example, from the rule lists in your example, to sort by cidr_ip would actually be: `rules.sort(key=lambda x: x.get('ip_ranges')[0].get('cidr_ip')))`, so this should be fixed, I just don't think this is doing what was intended either.

Reply:
@s-hertel 
Thanks for the response. I think this function "get_final_rules()" is trying to do normalization for the diff compare in main() function. 
So sorting in the "get_final_rules()" function is just trying to get the consistent format and order for the final_rule. Thus, sort by using the key in the final_rule is a way to make sure the consistence of the order. 
But in Python3, string cannot be compared, please refer to: https://docs.python.org/3/whatsnew/3.0.html#ordering-comparisons, and also not for the sorting method. 
Therefore, using the json.dump() and using sort_key with it is just trying to make sure the result is consistence, and I think that matches the function's goal.

Reply:
It is not sorting at all with this change. That's because the value returned from `json.dumps` (which is a string) is then used as the `key` parameter in the `sort` method. This doesn't error (but the key does not exist, so no sorting will occur), and will resolve the traceback in diff mode, but now it can show misleading diffs depending on the order the rules are given in the playbook.
For example, make a task with a list of different rules, and then make a following task that has the same rules, in a totally different order, plus one new rule. Then run the playbook with --diff; the output will show changes to a number of rules, not just the addition of one more.

Reply:
@s-hertel 
Oh, I got it, you're right. I originally use the json dump string to sort the final_rules, which is not correct. Sure thing, using json.loads(json.dump(sort_keys=True)) instead is more accurate for the purpose. Sorry for the misunderstanding, and thank you for the help.